articles/what-is-an-abi/index.md

```diff
@@ -0,0 +1,268 @@
+
+### Transferring Control
+
+First a little background on how the CPU executed a list of instructions. To know which instruction is needing to be executed there is a register called the instruction pointer, `rip`, that points to the next instruction needing to be executed. This will step through the code one instruction at a time.
+
+Back to the topic, in assembly code, we would call a function with the `call <callee addr>` instruction. This will do two things; First, it will push the current value of the instruction pointer on to the stack (this is the return address) and then it will change the value of the instruction pointer to be the address of the callee function.
+
+When we want to return from the callee and go back to the caller we can use the assembly instruction `ret` which will basically do the opposite of what `call` did. It will pop the return address off of the stack and into `rip` which will set the next instruction (what `rip` is pointing at) to be the instruction after the `call` in the caller code.
+
+Now you can think about how we could just keep calling function which would just keep pushing return addresses onto the stack until we choose to return which would pop all of the return addresses one at a time off of the stack until we are back to where we started.
+
+<!-- animation maybe, I don't really want to make one -->
+
+Calling and return from functions is important but it's more of a hardware thing then it as an ABI thing. So let's see how the ABI build off of this to allows for argument passing.
+
+### Saving and Restoring The Caller State
+
+Because the stack proved very useful in calling function and saving the return address we will use the stack again from saving the caller state.
+
+Let think about what ways we can do this. We could, before we call the callee, just back up every register's value we care about by pushing it onto the stack, and then after the callee is done we would just pop each value back into its register and carry on with the function. 
+
+This turns out to be very inefficient because if we have a lot of register in play and we call a function that doesn't use any of them we would end up filling the stack for no reason. It turns out that there is a better way to do this. It is the concept of caller-saved and callee-saved registers. If a register is marked as callee-saved and the callee wants to that register then they would have to back it up (push it onto the stack) before using it and restore it (pop it from the stack) before they returning.
+
+## Passing data to the callee and back
+
+Passing data is done in two different ways. The first is if you compile a program for 32 bits (x86) then all arguments are passed on the stack. The second way is if you compile for 64 bits (x64 architecture) then registers are used to pass arguments. If you have a lot of arguments then some are put onto the stack.

```

Comment:
The first is if you compile a program for 32 bits (x86) -> The first is when a program is compiled for 32 bits (x86)

Reply:
(x86) then all arguments are passed on the stack -> (x86); all arguments are passed on the stack

Reply:
he second way is if you compile for 64 bits (x64 architecture) -> The second is when a program is compiled for 64 bits (x64 architecture)

Reply:
(x64 architecture) then registers are used to pass arguments -> (x64 architecture); registers are used to pass arguments

Reply:
If you have a lot of arguments then some are put onto the stack. -> If there are too many arguments, the arguments that can't fit into a register will be put onto the stack.